File: agent/control/policy.py
from __future__ import annotations

from .models import ControlMode


CONTROL_TOOL_NAMES = {"ask_user", "propose_plan"}


class ControlPolicy:
    def __init__(self, manager):
        self.manager = manager

    def is_tool_allowed(self, name: str, registry) -> bool:
        if name == "ask_user":
            return True
        if name == "propose_plan":
            return self.manager.mode == ControlMode.PLAN.value
        if self.manager.mode != ControlMode.PLAN.value:
            return True
        tool = registry.get(name)
        if tool is None:
            return False
        return bool(getattr(tool, "read_only", False))

    def filtered_definitions(self, registry) -> list[dict]:
        definitions = registry.get_definitions()
        if self.manager.mode != ControlMode.PLAN.value:
            return [item for item in definitions if item.get("name") != "propose_plan"]
        return [item for item in definitions if self.is_tool_allowed(str(item.get("name") or ""), registry)]

File: packages/core/src/control/policy.ts
/**
 * ControlPolicy (MIG-CTRL-002)。对齐 Python `agent/control/policy.py`。
 * 经 PermissionPipeline.isToolExposed 暴露工具；plan 模式过滤 definitions。
 */
import { PermissionPipeline } from '../permissions/pipeline'
import type { ToolRegistry } from '../tools/registry'
import type { ToolDefinition } from '../tools/base'
import { ControlMode } from './models'

export const CONTROL_TOOL_NAMES = new Set(['ask_user', 'propose_plan'])

interface PolicyHost {
  readonly mode: string
}

export class ControlPolicy {
  private readonly manager: PolicyHost
  readonly permissionPipeline: PermissionPipeline

  constructor(manager: PolicyHost) {
    this.manager = manager
    this.permissionPipeline = new PermissionPipeline()
  }

  isToolAllowed(name: string, registry: ToolRegistry): boolean {
    return this.permissionPipeline.isToolExposed(name, this.manager.mode, { registry })
  }

  filteredDefinitions(registry: ToolRegistry): ToolDefinition[] {
    const definitions = registry.getDefinitions()
    if (this.manager.mode !== ControlMode.PLAN) {
      return definitions.filter((item) => item.name !== 'propose_plan')
    }
    // 已在计划模式：request_plan_mode 没有意义，与 propose_plan 的可见性互补
    return definitions.filter((item) => item.name !== 'request_plan_mode' && this.isToolAllowed(String(item.name ?? ''), registry))
  }
}
